### ISSUE TITLE
[JENKINS-32645] Rebuild Plugin triggering same job multiple times on slaves simultaneously

### ISSUE THREAD

**Issue by jenkins-infra-bot** [jenkinsci/nodelabelparameter-plugin#539]
<p><a href="https://wiki.jenkins-ci.org/display/JENKINS/Rebuild+Plugin" class="external-link" target="_blank" rel="nofollow noopener">https://wiki.jenkins-ci.org/display/JENKINS/Rebuild+Plugin</a></p>

<p>I have been using a really old version of Jenkins (Jenkins ver. 1.595) for our CI machines, (rebuilder 1.22) but recently decided to try and update Jenkins and the Plugins to the latest and greatest.</p>

<p>Since doing so, one of the issues I have encountered is when running "rebuild" on a build using all slaves, normally Jenkins would select the most appropriate slave, usually the last one used for that specific job (quite clever).. After updating Jenkins to 1.6.45 and then 1.6.46, and rebuild plugin 1.25</p>

<p>The duplicate jobs appear to have the specific slave names in, rather than before they were just referred to as "all", but despite calling "all" as the node, they have the slave name in the data in the queue.</p>

<p>Since I have 9 slaves, I end up with the same job running simultaneously 9 times.</p>

---
<details><summary><i>Originally reported by <img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/duncanbell.png" title="duncanbell's avatar" /> <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=duncanbell">duncanbell</a>, imported from: <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-32645" target="_blank">Rebuild Plugin triggering same job multiple times on slaves simultaneously</a></i></summary>
<i><ul>
<li><b>assignee</b>: <img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/amuniz.png" title="amuniz's avatar" /> <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=amuniz">amuniz</a>
<li><b>status</b>: Closed
<li><b>priority</b>: Major
<li><b>component(s)</b>: nodelabelparameter-plugin
<li><b>label(s)</b>: jenkins, plugin, rebuild, rebuild-last, slave, slave-agent, slaves, ssh-slave
<li><b>resolution</b>: Won't Fix
<li><b>resolved</b>: 2016-03-03T14:44:33+00:00
<li><b>votes</b>: 0
<li><b>watchers</b>: 2
<li><b>imported</b>: 20251210-235911
</ul></i>
<details><summary>Raw content of original issue</summary>

<pre>
<p><a href="https://wiki.jenkins-ci.org/display/JENKINS/Rebuild+Plugin" class="external-link" target="_blank" rel="nofollow noopener">https://wiki.jenkins-ci.org/display/JENKINS/Rebuild+Plugin</a></p>

<p>I have been using a really old version of Jenkins (Jenkins ver. 1.595) for our CI machines, (rebuilder 1.22) but recently decided to try and update Jenkins and the Plugins to the latest and greatest.</p>

<p>Since doing so, one of the issues I have encountered is when running "rebuild" on a build using all slaves, normally Jenkins would select the most appropriate slave, usually the last one used for that specific job (quite clever).. After updating Jenkins to 1.6.45 and then 1.6.46, and rebuild plugin 1.25</p>

<p>The duplicate jobs appear to have the specific slave names in, rather than before they were just referred to as "all", but despite calling "all" as the node, they have the slave name in the data in the queue.</p>

<p>Since I have 9 slaves, I end up with the same job running simultaneously 9 times.</p>
</pre>
</details>
</details>
<details><summary><i>environment</i></summary>

```
Jenkins 1.6.46 with rebuild 1.25, Ubuntu Trusty LTS, java version &quot;1.8.0_72&quot;<br/>
nodelabel parameter 1.71
```
</details>
<details><summary><i>1 attachment</i></summary>

- [jenkinsscreenshot.png](https://raw.githubusercontent.com/jenkinsci/attachments-from-jira-issues-last/refs/heads/main/attachments/31705/jenkinsscreenshot.png)
  > ![jenkinsscreenshot.png](https://raw.githubusercontent.com/jenkinsci/attachments-from-jira-issues-last/refs/heads/main/attachments/31705/jenkinsscreenshot.png)
</details>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/duncanbell.png" title="duncanbell's avatar" /> duncanbell:</i></summary>

<ul>
<li><i>Original <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-32645?focusedId=246934&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-246934">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>For testing I have copied over rebuild 1.22 on a Jenkins 1.6.46, and am still getting the same behaviour where clicking rebuild node=all triggers the same job on all my slaves, where as before it would just select one slave and run the job once.</p>

<p>It appears that in Jenkins 1.595, all jobs run on the slaves would retain the node label of "all", but with this new incorrect behaviour, they seem to have the node=slavename</p>

<p>I am going to check if using the old version of the ssh-slaves plugin (1.9) instead of 1.10 helps things.</p>
</pre>
</details>

</ul>
</details>

<p>For testing I have copied over rebuild 1.22 on a Jenkins 1.6.46, and am still getting the same behaviour where clicking rebuild node=all triggers the same job on all my slaves, where as before it would just select one slave and run the job once.</p>

<p>It appears that in Jenkins 1.595, all jobs run on the slaves would retain the node label of "all", but with this new incorrect behaviour, they seem to have the node=slavename</p>

<p>I am going to check if using the old version of the ssh-slaves plugin (1.9) instead of 1.10 helps things.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/duncanbell.png" title="duncanbell's avatar" /> duncanbell:</i></summary>

<ul>
<li><i>Original <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-32645?focusedId=246942&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-246942">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>Trying my older version of ssh-slaves plugin (1.9) has the same behaviour as 1.10.</p>

<p>The job starts off with...</p>

<p>(pending—Waiting for next available executor on all)</p>

<p>Then spawns to all slaves instead of just one...</p>
</pre>
</details>

</ul>
</details>

<p>Trying my older version of ssh-slaves plugin (1.9) has the same behaviour as 1.10.</p>

<p>The job starts off with...</p>

<p>(pending—Waiting for next available executor on all)</p>

<p>Then spawns to all slaves instead of just one...</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/duncanbell.png" title="duncanbell's avatar" /> duncanbell:</i></summary>

<ul>
<li><i>Original <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-32645?focusedId=249619&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-249619">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>Hi</p>

<p>I think my issue is actually with this plugin NodeLabel Parameter Plugin.. which has changed its behaviour...  I was using 1.5.1 and in 1.7 it changes to spawn to every slave...</p>

<p><a href="https://wiki.jenkins-ci.org/display/JENKINS/NodeLabel+Parameter+Plugin" class="external-link" target="_blank" rel="nofollow noopener">https://wiki.jenkins-ci.org/display/JENKINS/NodeLabel+Parameter+Plugin</a></p>

<p><a href="https://issues.jenkins-ci.org/browse/JENKINS-22185" class="external-link" target="_blank" rel="nofollow noopener">https://issues.jenkins-ci.org/browse/JENKINS-22185</a></p>

<p>Thanks</p>
</pre>
</details>

</ul>
</details>

<p>Hi</p>

<p>I think my issue is actually with this plugin NodeLabel Parameter Plugin.. which has changed its behaviour...  I was using 1.5.1 and in 1.7 it changes to spawn to every slave...</p>

<p><a href="https://wiki.jenkins-ci.org/display/JENKINS/NodeLabel+Parameter+Plugin" class="external-link" target="_blank" rel="nofollow noopener">https://wiki.jenkins-ci.org/display/JENKINS/NodeLabel+Parameter+Plugin</a></p>

<p><a href="https://issues.jenkins-ci.org/browse/JENKINS-22185" class="external-link" target="_blank" rel="nofollow noopener">https://issues.jenkins-ci.org/browse/JENKINS-22185</a></p>

<p>Thanks</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/amuniz.png" title="amuniz's avatar" /> amuniz:</i></summary>

<ul>
<li><i>Original <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-32645?focusedId=249623&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-249623">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>As I noted in <a href="https://issues.jenkins.io/browse/JENKINS-22185" title="Job with &quot;NodeLabel Parameter Plugin&quot; does not build in parallel on all nodes which share a label." class="issue-link" data-issue-key="JENKINS-22185"><del>JENKINS-22185</del></a>, you were relying in a bug so there is nothing to fix in this plugin. 
If you select "all" then the build will be scheduled to run in "all" nodes, nothing wrong. Unless you can provide an example of configuration where the expected behaviour is not what you get, I would close this issue as "Won't fix".</p>
</pre>
</details>

</ul>
</details>

<p>As I noted in <a href="https://issue-redirect.jenkins.io/issue/22185" title="Job with "NodeLabel Parameter Plugin" does not build in parallel on all nodes which share a label." class="issue-link" data-issue-key="<a class="jira-plain-text-key" href="https://issue-redirect.jenkins.io/issue/22185">JENKINS-22185</a>"><del>JENKINS-22185</del></a>, you were relying in a bug so there is nothing to fix in this plugin. <br/>
If you select "all" then the build will be scheduled to run in "all" nodes, nothing wrong. Unless you can provide an example of configuration where the expected behaviour is not what you get, I would close this issue as "Won't fix".</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/amuniz.png" title="amuniz's avatar" /> amuniz:</i></summary>

<ul>
<li><i>Original <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-32645?focusedId=249644&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-249644">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<blockquote><p>So there is no way for your plugin to random select one of N number of slaves to run the job on</p></blockquote>

<p>It's not <b>my</b> plugin. I think it's not possible (based on the <a href="https://wiki.jenkins-ci.org/display/JENKINS/NodeLabel+Parameter+Plugin" class="external-link" target="_blank" rel="nofollow noopener">documentation</a>). You can file it as a new feature (and propose a fix if you are able).</p>

<blockquote><p>I cannot see why someone would want to run the same job x 20 slaves</p></blockquote>

<p>It's a legitimate use case for others.</p>
</pre>
</details>

</ul>
</details>

<blockquote><p>So there is no way for your plugin to random select one of N number of slaves to run the job on</p></blockquote>

<p>It's not <b>my</b> plugin. I think it's not possible (based on the <a href="https://wiki.jenkins-ci.org/display/JENKINS/NodeLabel+Parameter+Plugin" class="external-link" target="_blank" rel="nofollow noopener">documentation</a>). You can file it as a new feature (and propose a fix if you are able).</p>

<blockquote><p>I cannot see why someone would want to run the same job x 20 slaves</p></blockquote>

<p>It's a legitimate use case for others.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/duncanbell.png" title="duncanbell's avatar" /> duncanbell:</i></summary>

<ul>
<li><i>Original <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-32645?focusedId=249648&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-249648">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>Thanks, I will look at raising a feature request to make the plugin work as it has done for the past zillion years.  Until then I can just deploy the working older version of the plugin I have.</p>
</pre>
</details>

</ul>
</details>

<p>Thanks, I will look at raising a feature request to make the plugin work as it has done for the past zillion years.  Until then I can just deploy the working older version of the plugin I have.</p>
